File: 0011-container-with-most-water/solution.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        int n = height.size();
        int l = 0, r = n - 1;
        int maxVol = INT_MIN;

        while(l < r) {
            maxVol = max(maxVol, (r - l) * min(height[l], height[r]));
            // We can do THREE things here
            // 1st:
            // if (height[l] >= height[r]) r--;
            // else l++;
            // 2nd:
            // if (height[l] < height[r]) l++;
            // else r--;
            // 3rd:
            if (height[l] <= height[r]) l++;
            else r--;
        }

        return maxVol;




        // int maxArea = 0;
        // int left = 0;
        // int right = height.size() - 1;

        // while (left < right) {
        //     maxArea = max(maxArea, (right - left) * min(height[left], height[right]));
        //     if (height[left] < height[right]) {
        //         left++;
        //     } 
        //     else {
        //         right--;
        //     }
        // }
        // return maxArea;        
    }
};

File: 0011-container-with-most-water/solution.java
class Solution {
    static {
        for (int i = 0; i < 65; i++) {
            maxArea(new int[] { 0, 0 });
        }}
    public static int maxArea(int[] height) {

        // Initialising two pointers left and right
        int left = 0;
        int right = height.length-1;

        // Initialising variable holding the value of MaxArea
        int maxArea = 0;

        while(left < right) {

            // Finding the min height of both pointers
            int min = Math.min(height[left],height[right]);


            // Comparing Current Area with MaxArea
            maxArea = Math.max(min*(right-left),maxArea);

            // Moving the small height pointer
           while(left<right && height[left] <= min) left++;
           while(left<right && height[right] <= min) right--;

        }

        // Return the maxArea Stored
        return maxArea;

    }
}
